core/src/main/java/hudson/FilePath.java

```diff
@@ -1600,14 +1605,22 @@ public Void invoke(File f, VirtualChannel channel) throws IOException {
     }
 
     /**
-     * Run chmod via jnr-posix
+     * Change permissions via NIO.
      */
     private static void _chmod(File f, int mask) throws IOException {
         // TODO WindowsPosix actually does something here (WindowsLibC._wchmod); should we let it?
         // Anyway the existing calls already skip this method if on Windows.
         if (File.pathSeparatorChar==';')  return; // noop
 
-        PosixAPI.jnr().chmod(f.getAbsolutePath(),mask);
+        if (Util.NATIVE_CHMOD_MODE) {
+            PosixAPI.jnr().chmod(f.getAbsolutePath(), mask);
+        } else {
+            try {
+                Files.setPosixFilePermissions(f.toPath().toAbsolutePath(), Util.modeToPermissions(mask));
+            } catch (InvalidPathException e) {
```

Comment:
I would propose to move it to a Restricted utility method, because it is going to be endless fun in this API.
Something like that: https://github.com/jenkinsci/remoting/blob/master/src/main/java/org/jenkinsci/remoting/util/PathUtils.java